Buffer TPs, TAs and TCs using the readoutlibs buffer implementation

philiprodrigues opened the pull request:
This PR changes the trigger object buffer implementation to use the one from readoutlibs, and adds buffering for TAs and TCs, in addition to TPs. It relies on the following two PRs:

https://github.com/DUNE-DAQ/detdataformats/pull/19
https://github.com/DUNE-DAQ/readoutlibs/pull/16

philiprodrigues committed:
Initial TPBuffer, using readoutlibs' buffer/data request interface

Doesn't compile yet

philiprodrigues committed:
Add boost::iostreams dependency

This is really a dependency of readoutlibs, not trigger, but I had to
put something here to make it build

philiprodrigues committed:
Actually implement TPWrapper::set_first_timestamp()

This turns out to be needed by the search functionality in DefaultRequestHandlerModel

philiprodrigues committed:
Implement do_work function

philiprodrigues committed:
Add schema file for configuration info

philiprodrigues committed:
Add TABuffer

Currently just a copy-paste of TPBuffer with appropriate
changes. Could consider doing this with templating

philiprodrigues committed:
Remove unused #includes

philiprodrigues committed:
Implement get_frame_iterator_timestamp for TPs

philiprodrigues committed:
Add schema for TABuffer

philiprodrigues committed:
Replace tpbuffer.jsonnet and tabuffer.jsonnet with txbuffer.jsonnet

An epiphany: the configs for TPBuffer and TABuffer are the same, so we
only need one of them

philiprodrigues committed:
Bugfix: serialize the TA and return serialized form in begin()/end()

This ensures that the TP list in the TA is also saved to the TR

philiprodrigues committed:
Add TCBuffer

philiprodrigues committed:
Remove unused get_timestamp() function

philiprodrigues committed:
Add TriggerCandidateNQ so we can send TCs over the network

This is currently needed for the hack that allows us to fan out TCs to
MLT and the TC buffer, in which we send TCs over a pub/sub network
connection from the trigger app back into the trigger app

philiprodrigues committed:
Make FakeTPCreatorHeartbeatMaker less verbose

philiprodrigues committed:
Add some debugging output

philiprodrigues committed:
Merge branch 'develop' into philiprodrigues/readoutlibs-buffer

eflumerf commented on CMakeLists.txt:
```diff
@@ -20,6 +20,7 @@ find_package(daqdataformats REQUIRED)
 find_package(detdataformats REQUIRED)
 find_package(detchannelmaps REQUIRED)
 find_package(hdf5libs REQUIRED)
+find_package(readoutlibs REQUIRED)
```
Is there an entry in cmake/triggerConfig.cmake.in?

philiprodrigues replied:
Oops, I forgot that. Thanks for catching it. Fixed now

philiprodrigues committed:
Add find_dependency(readoutlibs)

Thanks for Eric for pointing out the missing line

philiprodrigues commented:
Some more description of what to expect, motivated by questions from @bieryAtFnal :

* TCs are generated inside the trigger app for real and fake HSI triggers, so there should be at least one TC read out for every trigger record. (I need to check this actually works for HSI triggers).
* TP-based prescale triggers should also produce TAs and TCs that get buffered and read out. The number of TAs and TCs that end up in the trigger record will depend on the readout window (set by the TC) and the rate of TAs/TCs. There should be at least one of each, if I'm thinking this through correctly.
* https://github.com/DUNE-DAQ/daqconf/pull/36 describes how the region and element IDs are generated for storage in the HDF5 file

philiprodrigues committed:
Adapt to changes in readoutlibs for partial fragment on timeout

philiprodrigues committed:
Add print_ds_fragments test app

print_ds_fragments will print a text representation of the DS
fragments in an HDF5 file. The purpose is as a spot check that the
TP/TA/TC buffering and readout is working correctly.

philiprodrigues committed:
Merge branch 'develop' into philiprodrigues/return-partial-fragment-on-timeout to pick up iomanager-related changes

philiprodrigues committed:
Remove now-unused fragment output queue

philiprodrigues committed:
Replace QueueTimeoutExpired with TimeoutExpired

philiprodrigues committed:
Add Tee module, which duplicates its input on two output queues

This is useful for duplicating TP/TA/TC to the trigger chain and
readout buffers

philiprodrigues committed:
Make TXBuffer use new communications API

philiprodrigues committed:
Fix queue name

philiprodrigues committed:
Update an output warning

philiprodrigues committed:
Add get_name() to a TLOG() for benefit of debugging

philiprodrigues committed:
Merge branch 'develop' into philiprodrigues/return-partial-fragment-on-timeout

philiprodrigues committed:
Merge branch 'develop' into philiprodrigues/return-partial-fragment-on-timeout

philiprodrigues committed:
Merge branch 'develop' into philiprodrigues/return-partial-fragment-on-timeout

philiprodrigues committed:
Fix unused-variable compiler warnings

philiprodrigues committed:
When send fails, print _which_ queue the send was to

philiprodrigues committed:
In TXBuffer, make receive timeouts 0 so we can handle high rates

We were previously waiting on the receives from the TX input buffer
and on the data request input buffer. But then if the input rate from
one of queues (typically TPs) was greater than 1/(receive timeout),
the unconditional wait on the other queue would cause us to back
up. We now only wait if neither input queue received anything.
